array.rb

```diff
@@ -1,4 +1,58 @@
 class Array
+  # call-seq:
+  #   array.each {|element| ... } -> self
+  #   array.each -> Enumerator
+  #
+  # Iterates over array elements.
+  #
+  # When a block given, passes each successive array element to the block;
+  # returns +self+:
+  #
+  #   a = [:foo, 'bar', 2]
+  #   a.each {|element|  puts "#{element.class} #{element}" }
+  #
+  # Output:
+  #
+  #   Symbol foo
+  #   String bar
+  #   Integer 2
+  #
+  # Allows the array to be modified during iteration:
+  #
+  #   a = [:foo, 'bar', 2]
+  #   a.each {|element| puts element; a.clear if element.to_s.start_with?('b') }
+  #
+  # Output:
+  #
+  #   foo
+  #   bar
+  #
+  # When no block given, returns a new Enumerator:
+  #   a = [:foo, 'bar', 2]
+  #
+  #   e = a.each
+  #   e # => #<Enumerator: [:foo, "bar", 2]:each>
+  #   a1 = e.each {|element|  puts "#{element.class} #{element}" }
+  #
+  # Output:
+  #
+  #   Symbol foo
+  #   String bar
+  #   Integer 2
+  #
+  # Related: #each_index, #reverse_each.
+  def each
+    unless defined?(yield)
+      return to_enum(:each) { self.length }
+    end
+    _i = 0
+    value = nil
+    while Primitive.cexpr!(%q{ ary_fetch_next(self, LOCAL_PTR(_i), LOCAL_PTR(value)) })
+      yield value
+    end
```

Comment:
I'm fully in support of what is being done with this PR.

The part that I don't like so much is that `Primitive.cexpr` seems very mysterious to me. It's not really the case that a C expression is being compiled by a C compiler here, is it? I guess the syntax of a C function call is being parsed somewhere in the Ruby parser? Unless this whole thing is being compiled into a function call as a pre-processing step? 

The other concern I have is, passing pointers to CRuby stack slots. I'm not sure how we can handle that in YJIT. Could we come up with another solution? Originally you had proposed having an annotation that simply says a function can't be interrupted?

Reply:
> It's not really the case that a C expression is being compiled by a C compiler here, is it?

It is, if I understand your question correctly.

> I guess the syntax of a C function call is being parsed somewhere in the Ruby parser?

In Ruby's build process, `tool/mk_builtin_loader.rb` converts files like `array.rb` to serialized ISEQ binaries and a C source files like `array.rbinc`. `array.rbinc` contains something like this:

```c
#define LOCAL_PTR(local) local ## __ptr

static VALUE builtin_inline_class_50(struct rb_execution_context_struct *ec, const VALUE self)
{
    VALUE *const value__ptr = (VALUE *)&ec->cfp->ep[-3];
    MAYBE_UNUSED(const VALUE) value = *value__ptr;
    VALUE *const _i__ptr = (VALUE *)&ec->cfp->ep[-4];
    MAYBE_UNUSED(const VALUE) _i = *_i__ptr;

    return  ary_fetch_next(self, LOCAL_PTR(_i), LOCAL_PTR(value));
}
```

In the ISEQ binary for `Array#each`, `Primitive.cexpr!` is not a method call but `invokebuiltin` instruction that calls this C function `builtin_inline_class_50`.

So, `ary_fetch_next(self, LOCAL_PTR(_i), LOCAL_PTR(value))` is a valid C code, and is being compiled by a C compiler.

> The other concern I have is, passing pointers to CRuby stack slots. I'm not sure how we can handle that in YJIT. 

YJIT only needs to know an offset from EP for each local variable. ISEQ has a local table (`iseq->body->local_table`), and we can convert local variable names (`"_i"` and `"value"` in Rust) to EP offsets based on it.

> Could we come up with another solution? Originally you had proposed having an annotation that simply says a function can't be interrupted?

Right now, we have only ISEQ-level (i.e. not per-insn) annotations, and it's probably not a good idea to make a whole loop uninterruptible.

So we'd like to make a subset of instructions uninterruptible. But there's no easy way to encode that information in the instructions right now. `Primitive` (`invokebuiltin`) exists for doing something not possible in Ruby, so it's natural to use that for having a special behavior in part of the ISEQ.

Reply:
Ok, so the cexpr expressions get precompiled into C functions as part of a preprocessing step, which can then be called by CRuby.

> YJIT only needs to know an offset from EP for each local variable. ISEQ has a local table (iseq->body->local_table), and we can convert local variable names ("_i" and "value" in Rust) to EP offsets based on it.

That seems pretty hacky though. Not only do we need to parse and pattern-match on the cexpr primitives in YJIT, but we also need some way to expose EP offsets to a new primitive. The main issue that I see with this is that, as we get a more sophisticated register allocator, we want to be able to move away from the constraint that Ruby stack slots map to actual memory addresses. Local variables can live in a CPU register. If we have to force spill local variable slots so that we can refer to EP offsets, it adds more complexity to the JIT as well.

I think that your current solution works well in the context of the CRuby interpreter as it's implemented right now, but for the sake of the JIT, it would almost be better if we could introduce a new primitive that directly maps to a special YARV bytecode instruction. Or think of some other way to solve this problem.

Reply:
As we saw in a pairing session, I sketched out what we could do for this `Primitive` in YJIT https://github.com/k0kubun/ruby/pull/27. We can specialize `Primitive` of `Array#each` as if it was a C method. Since we can statically know the local variable offset for `_i` and `value`, it's no different from `getlocal` and `setlocal` to YJIT as long as we can specialize it that way.